packages/belos/src/BelosGmresPolyOp.hpp

```diff
@@ -606,23 +652,247 @@ namespace Belos {
 
+    if(info != 0){
+      std::cout << "GEEV solve : info = " << info << std::endl;
+    }
+
+    // Set index for sort function and sort Harmonic Ritz Values:
+    std::vector<int> index(dim_);
+    for(int i=0; i<dim_; ++i){ 
+      index[i] = i; 
+    }
+    SortModLeja(theta_,index);
+    //----End compute Harmonic Ritz Values (polynomial roots)------------
+
+    //-----------Section to add roots if needed for stability----------- 
+    
+    // Store theta (with cols for real and imag parts of Harmonic Ritz Vals) 
+    // as one vector of complex numbers to perform arithmetic:
+    std::vector<std::complex<MagnitudeType>> cmplxHRitz (dim_);
+    for(unsigned int i=0; i<cmplxHRitz.size(); ++i){
+      cmplxHRitz[i] = std::complex<MagnitudeType>( theta_(i,0), theta_(i,1) );
+    }
+   
+    // Compute product of factors (pof) to determine added roots: 
+    std::vector<MagnitudeType> pof (dim_,1.0);
+    for(int j=0; j<dim_; ++j) {
+      for(int i=0; i<dim_; ++i) {
+        if(i!=j) {
+          pof[j] = std::abs(pof[j]*(1.0-(cmplxHRitz[j]/cmplxHRitz[i])));
+        }
+      }
+    }
+
+    // Compute number of extra roots needed:
+    std::vector<int> extra (dim_);
+    int totalExtra = 0;
+    for(int i=0; i<dim_; ++i){
+      extra[i] = ceil((log10(pof[i])-4.0)/14.0);
+      if(extra[i] > 0){
+        totalExtra += extra[i];
+      }
+    }
+    if (totalExtra){
+      printer_->stream(Warnings) << "Warning: Need to add " << totalExtra << " extra roots." << std::endl;}
+
+    // If requested to add roots, append them to the theta matrix:
+    if(addRoots_ && totalExtra>0)
+    {
+      theta_.reshape(dim_+totalExtra,2);
+      // Make a matrix copy for perturbed roots:
+      Teuchos::SerialDenseMatrix<OT,MagnitudeType> thetaPert (Teuchos::Copy, theta_, dim_+totalExtra, 2);
+
+      //Add extra eigenvalues to matrix and perturb for sort:
+      int count = dim_;
+      for(int i=0; i<dim_; ++i){
+        for(int j=0; j< extra[i]; ++j){
+          theta_(count,0) = theta_(i,0);
+          theta_(count,1) = theta_(i,1);
+          thetaPert(count,0) = theta_(i,0)+(j+MCT::one())*5e-8;
+          thetaPert(count,1) = theta_(i,1);
+          ++count;
+        }
+      }
+
+      // Update polynomial degree:
+      dim_ += totalExtra;
+      if (totalExtra){
+        printer_->stream(Warnings) << "New poly degree is: " << dim_ << std::endl;}
+
+      // Create a new index and sort perturbed roots:
+      std::vector<int> index2(dim_);
+      for(int i=0; i<dim_; ++i){ 
+        index2[i] = i; 
+      }
+      SortModLeja(thetaPert,index2);
+      //Apply sorting to non-perterbed roots:
+      for(int i=0; i<dim_; ++i)
+      { 
+        thetaPert(i,0) = theta_(index2[i],0);
+        thetaPert(i,1) = theta_(index2[i],1);
+      }
+      theta_ = thetaPert;
+
+    } //----End root adding section------
+
+   }//End section for creating Roots poly 
```

Comment:
If you find yourself writing comments like "end root adding section," consider factoring out some code into a function.  It's hard to read really long function bodies.

Reply:
I factored this into a separate function.